### DISCUSSION TITLE
Postgres /bitnami/postgresql/data': Permission denied when following official helm chart guide on K3S

### DISCUSSION THREAD

**Discussion by SaarYogev** [apache/airflow#20481] [General]
### Official Helm Chart version

1.3.0 (latest released)

### Apache Airflow version

2.2.2 (latest released)

### Kubernetes Version

1.21 (K3s)

### Helm Chart configuration

```
kubectl create namespace airflow
helm repo add apache-airflow https://airflow.apache.org
helm install airflow apache-airflow/airflow --namespace airflow
```

### Docker Image customisations

No additional configuration

### What happened

Tried to use the starting guide:
https://airflow.apache.org/docs/helm-chart/stable/index.html
Finished this part:
https://airflow.apache.org/docs/helm-chart/stable/index.html#installing-the-chart
But the Airflow chart wasn't deployed properly.

Instead I got a CrashLoopBackOff for airflow-postgresql, with the following logs:
```
> kubectl logs airflow-postgresql-0 -n airflow
←[38;5;6mpostgresql ←[38;5;5m12:09:55.58 ←[0m
←[38;5;6mpostgresql ←[38;5;5m12:09:55.59 ←[0m←[1mWelcome to the Bitnami postgresql container←[0m
←[38;5;6mpostgresql ←[38;5;5m12:09:55.59 ←[0mSubscribe to project updates by watching ←[1mhttps://github.com/bitnami/bitnami-docker-postgresql←[0m
←[38;5;6mpostgresql ←[38;5;5m12:09:55.59 ←[0mSubmit issues and feature requests at ←[1mhttps://github.com/bitnami/bitnami-docker-postgresql/issues←[0m
←[38;5;6mpostgresql ←[38;5;5m12:09:55.59 ←[0m
←[38;5;6mpostgresql ←[38;5;5m12:09:55.62 ←[0m←[38;5;2mINFO ←[0m ==> ** Starting PostgreSQL setup **
←[38;5;6mpostgresql ←[38;5;5m12:09:55.64 ←[0m←[38;5;2mINFO ←[0m ==> Validating settings in POSTGRESQL_* env vars..
←[38;5;6mpostgresql ←[38;5;5m12:09:55.65 ←[0m←[38;5;2mINFO ←[0m ==> Loading custom pre-init scripts...
←[38;5;6mpostgresql ←[38;5;5m12:09:55.66 ←[0m←[38;5;2mINFO ←[0m ==> Initializing PostgreSQL database...
chmod: changing permissions of '/bitnami/postgresql/data': Operation not permitted
←[38;5;6mpostgresql ←[38;5;5m12:09:55.68 ←[0m←[38;5;3mWARN ←[0m ==> Lack of permissions on data directory!
chmod: changing permissions of '/bitnami/postgresql/data': Operation not permitted
←[38;5;6mpostgresql ←[38;5;5m12:09:55.68 ←[0m←[38;5;3mWARN ←[0m ==> Lack of permissions on data directory!
←[38;5;6mpostgresql ←[38;5;5m12:09:55.69 ←[0m←[38;5;2mINFO ←[0m ==> pg_hba.conf file not detected. Generating it...
←[38;5;6mpostgresql ←[38;5;5m12:09:55.69 ←[0m←[38;5;2mINFO ←[0m ==> Generating local authentication configuration
ls: cannot open directory '/bitnami/postgresql/data': Permission denied
```

### What you expected to happen

As the guide says:
`The command deploys Airflow on the Kubernetes cluster in the default configuration`

### How to reproduce

Use the guide to reproduce:
https://airflow.apache.org/docs/helm-chart/stable/index.html
It appears every time.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
This seems to be specific to K3s - which apparently has problems with postrgresql subchart. The subchart comes from BitNami, so you should raise the issue there.

https://github.com/bitnami/charts/tree/master/bitnami/postgresql

I thnk it would be great if you reproduce this issue only with the bitnami subchart and raise it directly with them - if you provide aa reproducible case for them, they are likely to be able to fix it quickly.

The PostgreSQL subchart is really only for convenience/local development and you should really setup an external database for the Airflow Metadata DB: https://airflow.apache.org/docs/helm-chart/stable/production-guide.html#database 

I think it is unlikely we will spend too much time on this specific setting with K3s, so your best bet is to deal directly with bitnami on that.

**Comment by SaarYogev** [comment]
Thanks for the fast reply!

Unfortunately, the issue actually isn't with Bitnami, as I can deploy their Postgresql easily with Helm, with the default settings, on the same K3s cluster.
So this issue needs to reopened to better look into this.

Also, out of curiosity, why do you want an external cluster for production? Postgres works nicely in K8s production environments: https://dok.community/blog/deploying-postgres-on-kubernetes/
But I do see this request to use external database for production with a lot of products so I assume there's a good reason for it.
Why is that the case here?

**Comment by potiuk** [comment]
> Unfortunately, the issue actually isn't with Bitnami, as I can deploy their Postgresql easily with Helm, with the default settings, on the same K3s cluster. So this issue needs to reopened to better look into this.

TCan you try on a different cluster that K3s ?  Because if it is K3S. then I am afraid this at most can be turned into discussion. This is not a bug, really if the internal Postgres does not work with K3s - it's at most an inconvenence -  possibly this could be addressed to K3S then. 

> Also, out of curiosity, why do you want an external cluster for production? Postgres works nicely in K8s production environments: https://dok.community/blog/deploying-postgres-on-kubernetes/

The fact that you can do something does not mean you should. If you watched that presentation, you will likely see all the hoops that you have to go through as K8S adrministrator. They say about days of work to make the deployment robust. So this is not something simple bitnami chart will handle.

> But I do see this request to use external database for production with a lot of products so I assume there's a good reason for it.
> Why is that the case here?

It's everywhere not here. K8S is simply not (yet) fully ready for "stateful" components. It was not designed to do it initially and a lot of changes in it were implemented to get closer to be able to support them. But it's simply not best tool to do it and you have to be an expert in K8S and expert in DB administration and expert in your own deployment to make it really robust. You can. Sure. Should you ? Maybe, if your job is paid by hour, probably it's even a good idea for you. Is it good for the business that uses Airlfow ? Probably not. using a managed database service is likely going to be cheaper and more stable overall.
